04_Scheduling/03_Processes_And_Threads.md

```diff
@@ -169,94 +169,165 @@ Now any further operations on this file can use the returned id to reference thi
-* They can have different statuses just like processes, they can be running, waiting for their turn, sleeping, etc, it again depends on design choices, but we need a status field too.
-* Another optional information to make thread identification easier can be a thread name.
+```c
+typedef struct {
+    size_t pid;
+    thread_t* threads;
+    void* root_page_table;
+    char name[NAME_MAX_LEN];
+} process_t;
+```
 
-We can wrap the information above in a new data structure, and update the process structure accordingly: 
+We're going to use a linked list as our data structure to manage threads. Adding a new thread would look something like the following:
 
 ```c
-struct {
-    size_t tid;
-    cpu_status_t* context;
-    void* stack;
-    thread_status_t status;
-    char *name[THREAD_NAME_LEN];
-} thread_t
+size_t next_thread_id = 0;
+
+thread_t* add_thread(process_t* proc, char* name, void(*function)(void*), void* arg)
+{
+    thread_t* thread = malloc(sizeof(thread_t));
+    if (proc->threads = NULL)
+        proc->threads = thread;
+    else
+    {
+        for (thread_t* scan = proc->threads; scan != NULL; scan = scan->next)
+        {
+            if (scan->next != NULL)
+                continue;
+            scan->next = thread;
+            break;
+        }
+    }
+
+    strncpy(thread->name, name, NAME_MAX_LEN);
+    thread->tid = next_thread_id++;
+    thread->status = READY;
+    thread->next = NULL:
+    thread->context.iret_ss = KERNEL_SS;
+    thread->context.iret_rsp = alloc_stack();
+    thread->context.iret_flags = 0x202;
+    thread->context.iret_cs = KERNEL_CS;
+    thread->context.iret_rip = (uint64_t)function;
+    thread->context.rdi = (uint64_t)arg;
+    thread->context.rbp = 0;
+
+    return thread;
+}
 ```
 
-The `thread_status_t` field is not defined yet, so depending on the design decisions and or the scheduling algorithm implemented it can be the same as the process statuses or not, for now let's assume that the thread statuses are the same with process statuses, and just declare it as a new data type: 
+You'll notice this function looks almost identical to the `create_process` function from before. That's because a lot of it is the same! The first part of the function is just inserting the new thread at the end of the list of threads.
+
+Let's look at how our `create_process` function would look now:
 
 ```c
-typedef status_t thread_status_t;
+process_t* create_process(char* name)
+{
+    process_t* process = alloc_process();
+    process->pid = next_process_id++;
+    process->threads = NULL;
+    process->root_page_table = vmm_create();
+    strncpy(process->name, name, NAME_MAX_LEN);
+    return process;
+}
 ```
 
-We need to update also the `process_t` data structure with few changes: 
+The `alloc_process` function is just a placeholder, for brevity, `vmm_create` is also a placeholder. Replace these with your own code.
```

Comment:
even if iti quite clear just spending a sentence to explain what are the two function expected output, especially for the vmm_create can be useful?